### PR TITLE
GEOMESA-705 Don't encode null attributes in the index

### PR DESCRIPTION
**Pull request by elahrvivaz**

### CODE REVIEW

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/tables/AttributeTable.scala [hunk 1]
```diff
@@ -164,20 +167,22 @@ object AttributeTable extends GeoMesaTable with Logging {
...
-  def encode(valueOption: Option[Any], descriptor: AttributeDescriptor): Seq[String] = {
-    val value = valueOption.getOrElse(nullString)
-    if (descriptor.isCollection) {
+  def encode(value: Any, descriptor: AttributeDescriptor): Seq[String] = {
+    if (value == null) {
+      Seq.empty
+    } else if (descriptor.isCollection) {
       // encode each value into a separate row
-      value.asInstanceOf[JCollection[_]].asScala.toSeq.map(Option(_).getOrElse(nullString)).map(typeEncode)
+      value.asInstanceOf[JCollection[_]].toSeq.flatMap(Option(_).map(typeEncode).filterNot(_.isEmpty))
```
Comment (apc3n) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/tables/AttributeTable.scala]: what about filter(_.nonEmpty)

Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/tables/AttributeTable.scala]: I feel like there is some minor cost involved in the implicit conversion to a GenTraversableOnce with .nonEmpty, but I can't really confirm that... Any thoughts?

Comment (apc3n) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/tables/AttributeTable.scala]: oh yeah gotcha. I didn't see that it was a JCollection. Yeah avoid the conversion and leave as it

Comment (apc3n) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/tables/AttributeTable.scala]: nvm again... yeah I see that it is a string.. It's fine the way it is

**Comment by jdk2pq** [comment]
Read through this -- looks good. Built, and all went as expected.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/process/unique/UniqueProcess.scala [hunk 2]
```diff
@@ -233,13 +233,11 @@ object AttributeVisitor {
...
   /**
    * Returns a filter that is equivalent to Filter.INCLUDE, but against the attribute index.
-   * Excludes null values, since we don't care about those anyway.
    *
    * @param attribute
    * @return
    */
-  def getIncludeAttributeFilter(attribute: String) =
-    ff.greater(ff.property(attribute), ff.literal(AttributeTable.nullString))
+  def getIncludeAttributeFilter(attribute: String) = ff.greater(ff.property(attribute), ff.literal(""))
```
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/process/unique/UniqueProcess.scala]: Quick question: is "" the 'least' literal/string/lexicoded value?

Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/process/unique/UniqueProcess.scala]: yeah, it ends up acting as a 'prefix' function, e.g. if the rows are:
<sft><attr-name><encoded value>
we create
<sft><attr-name>
which will sort before any encoded value (empty strings are handled as prefixes on AttributeTable#202 and then AttributeIdxStrategy#379)

**Comment by jnh5y** [comment]
Merged as 8af364710c120eb3d979a164b41f5f2964c8caf0 on 1.5.

**Comment by jnh5y** [comment]
Actually, d6882306cccde71a4b011f0ba8253a717f478bb3.

**Pull request closed without merging**